# **Personal Finance** - Analyze Financial Habits

# Sobre el conjunto de datos

Usabilidad de la base de datos: 7.06

Tamaño base de datos: 4.19 MB

### **Contexto**

En un mundo donde las decisiones financieras personales impactan directamente en la estabilidad económica de los hogares, comprender los patrones de comportamiento financiero es fundamental. Sin embargo, los datos reales sobre finanzas personales suelen ser sensibles y difíciles de obtener debido a su naturaleza privada.

El Personal Finance ML Dataset ha sido creado para simular de forma sintética (2020 - 2024), pero realista, el comportamiento financiero de personas de diferentes regiones, niveles de ingreso y situaciones crediticias. Este conjunto de datos reproduce de manera real las dinámicas del mundo real, permitiendo analizar hábitos financieros y estudiar la relación entre ingresos, gastos y deudas.

### **Descripción de los datos**

El dataset contiene 32.424 registros individuales (filas), donde cada fila representa el perfil financiero de una persona. Los datos incluyen información demográfica, económica, crediticia y temporal, con el objetivo de reflejar la diversidad de comportamientos financieros a nivel global.

* **user_id**: Identificador único que representa a cada individuo.
* **age**: Edad del individuo, expresada en años (rango de 18 a 70).
* **gender**: Género del individuo: Masculino, Femenino u Otro.
* **education_level**: Nivel educativo más alto alcanzado (por ejemplo, Secundaria, Grado, Máster o Doctorado).
* **employment_status**: Situación laboral actual del individuo (por ejemplo, Empleado, Desempleado, Estudiante, Jubilado).
* **job_title**: Puesto de trabajo o cargo desempeñado.
* **monthly_income_usd**: Ingreso mensual aproximado en dólares estadounidenses.
* **monthly_expenses_usd**: Gastos mensuales aproximados en dólares estadounidenses.
* **savings_usd**: Total de ahorros acumulados por el individuo.
* **has_loan**: Indica si el individuo tiene un préstamo activo: Sí o No.
* **loan_type**: Tipo de préstamo asociado, en caso de tenerlo (por ejemplo, personal, hipotecario, automotriz o educativo).
* **loan_amount_usd**: Monto total del préstamo en dólares estadounidenses.
* **loan_term_months**: Duración del préstamo expresada en meses.
* **monthly_emi_usd**: Cuota mensual del préstamo (EMI, Equated Monthly Installment).
* **loan_interest_rate_pct**: Tasa de interés anual aplicada al préstamo, expresada en porcentaje.
* **debt_to_income_ratio**: Relación entre los pagos de deuda y los ingresos del individuo (DTI).
* **credit_score**: Puntuación crediticia sintética en una escala de 300 a 850.
* **savings_to_income_ratio**: Relación entre los ahorros y los ingresos anuales del individuo.
* **region**: Región geográfica en la que reside el individuo.
* **record_date**: Fecha de registro de la información financiera.
### **Objetivo del Proyecto**

El objetivo de este proyecto es realizar un proceso completo de limpieza y análisis exploratorio (EDA) sobre el conjunto de datos del Personal Finance ML Dataset, con el fin de comprender las características principales de la población representada y extraer patrones significativos sobre el comportamiento financiero individual.

De manera provisional, el análisis podría orientarse hacia alguno de los siguientes enfoques:
	•	Evaluar la salud financiera de los individuos según sus ingresos, gastos y nivel de ahorro.
	•	Identificar patrones de riesgo crediticio mediante el estudio del credit_score, debt_to_income_ratio y loan_status.
	•	Segregar perfiles financieros en función de la región, nivel educativo o tipo de préstamo.
	•	Explorar relaciones entre ingresos, ahorros y endeudamiento, generando visualizaciones que faciliten la interpretación.

En una segunda etapa, los resultados del EDA podrán servir como base para desarrollar un modelo de predicción o segmentación, dependiendo del enfoque que finalmente se elija.

# 1.-Librerías

In [1]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd #manejo de estructura de datos tabulares

# Visualización de datos
# -----------------------------------------------------------------------
import matplotlib.pyplot as plt
import seaborn as sns


# importamos los warnings
# -----------------------------------------------------------------------
import warnings
warnings.filterwarnings('ignore')

# importamos configuracion de pandas
# -----------------------------------------------------------------------
pd.options.display.max_columns = None

# 2.-Cargar datos

In [ ]:
# Cargamos los datos desde la carpeta data/raw
df = pd.read_csv("../data/raw/synthetic_personal_finance_dataset.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32424 entries, 0 to 32423
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   user_id                  32424 non-null  object 
 1   age                      32424 non-null  int64  
 2   gender                   32424 non-null  object 
 3   education_level          32424 non-null  object 
 4   employment_status        32424 non-null  object 
 5   job_title                32424 non-null  object 
 6   monthly_income_usd       32424 non-null  float64
 7   monthly_expenses_usd     32424 non-null  float64
 8   savings_usd              32424 non-null  float64
 9   has_loan                 32424 non-null  object 
 10  loan_type                12995 non-null  object 
 11  loan_amount_usd          32424 non-null  float64
 12  loan_term_months         32424 non-null  int64  
 13  monthly_emi_usd          32424 non-null  float64
 14  loan_interest_rate_pct

In [3]:
df.head()

user_id  age  gender education_level employment_status    job_title  \
0  U00001   56  Female     High School     Self-employed  Salesperson   
1  U00002   19  Female             PhD          Employed  Salesperson   
2  U00003   20  Female          Master          Employed      Teacher   
3  U00004   25    Male             PhD          Employed      Manager   
4  U00005   53  Female             PhD          Employed      Student   

   monthly_income_usd  monthly_expenses_usd  savings_usd has_loan  loan_type  \
0             3531.69               1182.59    367655.03       No        NaN   
1             3531.73               2367.99    260869.10      Yes  Education   
2             2799.49               1003.91    230921.21       No        NaN   
3             5894.88               4440.12    304815.51      Yes   Business   
4             5128.93               4137.61    461509.48       No        NaN   

   loan_amount_usd  loan_term_months  monthly_emi_usd  loan_interest_rate_pct  \
0             0.00                 0             0.00                    0.00   
1        146323.34                36          4953.50                   13.33   
2             0.00                 0             0.00                    0.00   
3         93242.37                24          4926.57                   23.93   
4             0.00                 0             0.00                    0.00   

   debt_to_income_ratio  credit_score  savings_to_income_ratio         region  \
0                  0.00           430                     8.68          Other   
1                  1.40           543                     6.16  North America   
2                  0.00           754                     6.87         Africa   
3                  0.84           461                     4.31         Europe   
4                  0.00           516                     7.50         Africa   

  record_date  
0  2024-01-09  
1  2022-02-13  
2  2022-05-12  
3  2023-10-02  
4  2021-08-07

Con esto ya tenemos un resumen del DataFrame: tiene 32424 filas y 20 columnas. Vemos el nombre de las columnas y también el Dtype. Las columnas contienen varias columnas object que seguramente podremos tratar más adelante, por ejemplo, conviertiéndolas en boleanas.

# 3.-Exploración inicial de los datos

## 3.1.-Nulos y tipos de datos

Saber cuántos valores no nulos tenemos es clave para analizar los datos y detectar posibles líos, como valores que faltan o que no cuadran. Eso sí, si en lugar de fijarnos en los no nulos miramos directamente los nulos, el concepto resulta más sencillo de entender.

In [4]:
# creamos un dataframe con las columnas que necesitamos
df_resume = pd.DataFrame(columns=["count", "null", "null_%", "dtype", "unique"])

dict_resume = {}
for i in df.columns:
    dict_resume["count"] = df[i].count()
    dict_resume["null"] = df[i].isnull().sum()
    dict_resume["null_%"] = round(df[i].isnull().sum()/len(df),2)
    dict_resume["dtype"] = df[i].dtype
    dict_resume["unique"] = df[i].nunique()
    df_aux = pd.DataFrame(dict_resume, index=[i])
    df_resume = pd.concat([df_resume, df_aux], axis=0)

df_resume

count   null  null_%    dtype unique
user_id                  32424      0     0.0   object  32424
age                      32424      0     0.0    int64     52
gender                   32424      0     0.0   object      3
education_level          32424      0     0.0   object      5
employment_status        32424      0     0.0   object      4
job_title                32424      0     0.0   object      9
monthly_income_usd       32424      0     0.0  float64  30449
monthly_expenses_usd     32424      0     0.0  float64  31319
savings_usd              32424      0     0.0  float64  32415
has_loan                 32424      0     0.0   object      2
loan_type                12995  19429     0.6   object      4
loan_amount_usd          32424      0     0.0  float64  12993
loan_term_months         32424      0     0.0    int64     10
monthly_emi_usd          32424      0     0.0  float64  12948
loan_interest_rate_pct   32424      0     0.0  float64   2684
debt_to_income_ratio     32424      0     0.0  float64   1541
credit_score             32424      0     0.0    int64    551
savings_to_income_ratio  32424      0     0.0  float64    991
region                   32424      0     0.0   object      5
record_date              32424      0     0.0   object   1461

In [5]:
# lo pasamos a funcion
def data_resume(df):
    """
    Devuelve un resumen de los datos del dataframe con el recuento de valores, valores nulos, porcentaje de valores nulos, tipo de dato y número de valores únicos por columna. 

    Args:
        df (pd.DataFrame): dataframe a resumir.

    Returns:
        pd.DataFrame: resumen de los datos del dataframe.
    """
    df_resume = pd.DataFrame(columns=["count", "null", "null_%", "dtype", "unique"])
    dict_resume = {}
    for i in df.columns:
        dict_resume["count"] = df[i].count()
        dict_resume["null"] = df[i].isnull().sum()
        dict_resume["null_%"] = round(df[i].isnull().sum()/len(df),2)
        dict_resume["dtype"] = df[i].dtype
        dict_resume["unique"] = df[i].nunique()
        df_aux = pd.DataFrame(dict_resume, index=[i])
        df_resume = pd.concat([df_resume, df_aux], axis=0)
    return df_resume.sort_values(by="null_%", ascending=False)

In [6]:
data_resume(df)

count   null  null_%    dtype unique
loan_type                12995  19429     0.6   object      4
age                      32424      0     0.0    int64     52
region                   32424      0     0.0   object      5
savings_to_income_ratio  32424      0     0.0  float64    991
credit_score             32424      0     0.0    int64    551
debt_to_income_ratio     32424      0     0.0  float64   1541
loan_interest_rate_pct   32424      0     0.0  float64   2684
monthly_emi_usd          32424      0     0.0  float64  12948
loan_term_months         32424      0     0.0    int64     10
loan_amount_usd          32424      0     0.0  float64  12993
user_id                  32424      0     0.0   object  32424
has_loan                 32424      0     0.0   object      2
savings_usd              32424      0     0.0  float64  32415
monthly_expenses_usd     32424      0     0.0  float64  31319
monthly_income_usd       32424      0     0.0  float64  30449
job_title                32424      0     0.0   object      9
employment_status        32424      0     0.0   object      4
education_level          32424      0     0.0   object      5
gender                   32424      0     0.0   object      3
record_date              32424      0     0.0   object   1461